go/marshal/decode_test.go

```diff
@@ -0,0 +1,385 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package marshal
+
+import (
+	"bytes"
+	"fmt"
+	"math"
+	"reflect"
+	"testing"
+
+	"github.com/attic-labs/noms/go/types"
+	"github.com/attic-labs/testify/assert"
+)
+
+func TestDecode(tt *testing.T) {
+	assert := assert.New(tt)
+
+	t := func(v types.Value, ptr interface{}, expected interface{}) {
+		p := reflect.ValueOf(ptr)
+		assert.Equal(reflect.Ptr, p.Type().Kind())
+		err := Unmarshal(v, p.Interface())
+		assert.NoError(err)
+		assert.Equal(expected, p.Elem().Interface())
+
+		// Also test that types.Value is passed through
+		var v2 types.Value
+		err = Unmarshal(v, &v2)
+		assert.NoError(err)
+		assert.True(v.Equals(v2))
+	}
+
+	for _, n := range []float32{0, 42, 3.14159265359, math.MaxFloat32} {
+		var f32 float32
+		t(types.Number(n), &f32, float32(n))
+	}
+
+	for _, n := range []float64{0, 42, 3.14159265359, math.MaxFloat64} {
+		var f64 float64
+		t(types.Number(n), &f64, float64(n))
+	}
+
+	for _, n := range []int8{0, 42, math.MaxInt8} {
+		var i8 int8
+		t(types.Number(n), &i8, int8(n))
+	}
+
+	for _, n := range []int16{0, 42, math.MaxInt16} {
+		var i16 int16
+		t(types.Number(n), &i16, int16(n))
+	}
+
+	for _, n := range []int32{0, 42, math.MaxInt32} {
+		var i32 int32
+		t(types.Number(n), &i32, int32(n))
+	}
+
+	// int is at least int32
+	for _, n := range []int{0, 42, math.MaxInt32} {
+		var i int
+		t(types.Number(n), &i, int(n))
+	}
+
+	// There is precision loss for values above Math.pow(2, 53) - 1
+	for _, n := range []int64{0, 42, int64(math.Pow(2, 53) - 1)} {
+		var i64 int64
+		t(types.Number(n), &i64, int64(n))
+	}
+
+	for _, n := range []uint8{0, 42, math.MaxUint8} {
+		var ui8 uint8
+		t(types.Number(n), &ui8, uint8(n))
+	}
+
+	for _, n := range []uint16{0, 42, math.MaxUint16} {
+		var ui16 uint16
+		t(types.Number(n), &ui16, uint16(n))
+	}
+
+	for _, n := range []uint32{0, 42, math.MaxInt32} {
+		var ui32 uint32
+		t(types.Number(n), &ui32, uint32(n))
+	}
+
+	// uint is at least uint32
+	for _, n := range []uint{0, 42, math.MaxInt32} {
+		var ui uint
+		t(types.Number(n), &ui, uint(n))
+	}
+
+	// There is precision loss for values above Math.pow(2, 53) - 1
+	for _, n := range []uint64{0, 42, uint64(math.Pow(2, 53) - 1)} {
+		var ui64 uint64
+		t(types.Number(n), &ui64, uint64(n))
+	}
+
+	var b bool
+	t(types.Bool(true), &b, true)
+	t(types.Bool(false), &b, false)
+
+	for _, s := range []string{"", "s", "hello", "💩"} {
+		var s2 string
+		t(types.String(s), &s2, s)
+	}
+
+	var list types.List
+	list2 := types.NewList()
```

Comment:
In all these tests with Noms collections, I would put something in e.g. `list2`, to make sure that the target isn't just getting an empty collection stuffed in there, but rather getting the appropriate Value assigned.